src/nekrs_driver.cpp

```diff
@@ -0,0 +1,229 @@
+    n_gll_ = (poly_deg_ + 1) * (poly_deg_ + 1) * (poly_deg_ + 1);
+
+    x_ = nekrs::mesh::x();
+    y_ = nekrs::mesh::y();
+    z_ = nekrs::mesh::z();
+    element_info_ = nekrs::mesh::elementInfo();
+
+    // rho energy is field 1 (0-based) of rho
+    rho_cp_ = nekrs::rhoCp();
+
+    temperature_ = nekrs::scalarFields();
+
+    // Construct lumped mass matrix from vgeo
+    // See cdsSetup in vendor/nekRS/src/core/insSetup.cpp
+    mass_matrix_.resize(n_local_elem_ * n_gll_);
+    for (gsl::index e = 0; e < n_local_elem_; ++e) {
+      for (gsl::index n = 0; n < n_gll_; ++n) {
+        mass_matrix_[e * n_gll_ + n] =
+          nekrs::mesh::vgeo()[e * n_gll_ * nekrs::mesh::nVgeo() + JWID * n_gll_ + n];
+      }
+    }
+
+    init_displs();
+  }
+  comm_.Barrier();
+}
+
+void NekRSDriver::init_step()
+{
+  auto min = std::min_element(localq_->cbegin(), localq_->cend());
+  auto max = std::max_element(localq_->cbegin(), localq_->cend());
+}
+
+void NekRSDriver::solve_step()
+{
+  const auto start_time = nekrs::startTime();
+  const auto final_time = nekrs::finalTime();
+  const auto dt = nekrs::dt();
+
+  time_ = start_time;
+  tstep_ = 1;
+
+  while ((final_time - time_) / (final_time * dt) > 1e-6) {
+    nekrs::runStep(time_, dt, tstep_);
+    time_ += dt;
+    nekrs::udfExecuteStep(time_, tstep_, 0);
+    ++tstep_;
+  }
+  // LOOK HERE!!!
+  nekrs::copyToNek(time_, tstep_);
+}
+
+void NekRSDriver::write_step(int timestep, int iteration)
+{
+  nekrs::copyToNek(timestep, iteration);
+  nekrs::nekOutfld();
+  return;
+}
+
+Position NekRSDriver::centroid_at(int32_t local_elem) const
+{
+  Expects(local_elem < n_local_elem());
+  Position c{0., 0., 0.};
+  double mass = 0.;
+  for (gsl::index i = 0; i < n_gll_; ++i) {
+    auto idx = local_elem * n_gll_ + i;
+    c.x += x_[idx] * mass_matrix_[idx];
+    c.y += y_[idx] * mass_matrix_[idx];
+    c.z += z_[idx] * mass_matrix_[idx];
+    mass += mass_matrix_[idx];
+  }
+  c.x /= mass;
+  c.y /= mass;
+  c.z /= mass;
+  return c;
+}
+
+std::vector<Position> NekRSDriver::centroid_local() const
+{
+  std::vector<Position> c(n_local_elem());
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    c[i] = this->centroid_at(i);
+  }
+  return c;
+}
+
+double NekRSDriver::volume_at(int32_t local_elem) const
+{
+  Expects(local_elem < n_local_elem());
+  double v = 0.;
+  for (int32_t i = 0; i < n_gll_; ++i) {
+    v += mass_matrix_[local_elem * n_gll_ + i];
+  }
+  return v;
+}
+
+std::vector<double> NekRSDriver::volume_local() const
+{
+  std::vector<double> v(n_local_elem());
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    v[i] = this->volume_at(i);
+  }
+  return v;
+}
+
+double NekRSDriver::temperature_at(int32_t local_elem) const
+{
+  Expects(local_elem < n_local_elem());
+
+  double sum0 = 0.;
+  double sum1 = 0.;
+  for (int32_t i = 0; i < n_gll_; ++i) {
+    auto idx = local_elem * n_gll_ + i;
+    sum0 += rho_cp_[idx] * temperature_[idx];
+    sum1 += rho_cp_[idx];
```

Comment:
I forget -- what was the conclusion about how we _should_ be doing the averaging?

Reply:
This is what we do in Nek5000, as well.  Sorry, but I forget what the alternate proposal was.

Reply:
The only piece I remember was the argument that it probably won't make much of a difference, so sticking with the status quo seems pretty acceptable :+1: